cli/command_repository_upgrade.go

```diff
@@ -0,0 +1,291 @@
+	"path/filepath"
+	"time"
+
+	"github.com/alecthomas/kingpin"
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/internal/epoch"
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/content"
+)
+
+type commandRepositoryUpgrade struct {
+	forceRollback bool
+	skip          bool
+	force         bool
+
+	// lock settings
+	advanceNoticeDuration time.Duration
+	ioDrainTimeout        time.Duration
+	statusPollInterval    time.Duration
+
+	svc advancedAppServices
+}
+
+const (
+	experimentalWarning = `WARNING: The upgrade command is an EXPERIMENTAL feature. Please use with caution.
+
+You will need to set the env variable KOPIA_UPGRADE_LOCK_ENABLED in order to use this feature.
+`
+	upgradeLockFeatureEnv                = "KOPIA_UPGRADE_LOCK_ENABLED"
+	maxPermittedClockDrift time.Duration = 5 * time.Minute
+)
+
+// MaxPermittedClockDrift is overridable interface for tests to define their
+// own constants so that they do not have to wait for the default clock-drift to
+// settle.
+//
+// nolint:gochecknoglobals
+var MaxPermittedClockDrift = func() time.Duration { return maxPermittedClockDrift }
+
+func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent commandParent) {
+	// override the parent, the upgrade sub-command becomes the new parent here-onwards
+	parent = parent.Command("upgrade", fmt.Sprintf("Upgrade repository format.\n\n%s", warningColor.Sprint(experimentalWarning))).Hidden().
+		Validate(func(tmpCmd *kingpin.CmdClause) error {
+			if v := os.Getenv(c.svc.EnvName(upgradeLockFeatureEnv)); v == "" {
+				return errors.Errorf("please set %q env variable to use this feature", upgradeLockFeatureEnv)
+			}
+			return nil
+		})
+
+	beginCmd := parent.Command("begin", "Begin upgrade.").Default()
+	beginCmd.Flag("advance-notice", "Advance notice for upgrade to allow enough time for other Kopia clients to notice the lock").DurationVar(&c.advanceNoticeDuration)
+	beginCmd.Flag("io-drain-timeout", "Max time it should take all other Kopia clients to drop repository connections").Default(repo.DefaultRepositoryBlobCacheDuration.String()).DurationVar(&c.ioDrainTimeout)
+	beginCmd.Flag("force", "Force using an unsafe io-drain-timeout").Default("false").Hidden().BoolVar(&c.force)
+	beginCmd.Flag("status-poll-interval", "An advisory polling interval to check for the status of upgrade").Default("60s").DurationVar(&c.statusPollInterval)
+
+	// upgrade phases
+
+	// Set the upgrade lock intent.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.setLockIntent)))
+	// If requested then drain all the clients otherwise stop here.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.drainOrCommit)))
+	// If the lock is fully established then perform the upgrade.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.upgrade)))
+	// Commit the upgrade and revoke the lock, this will also cleanup any
+	// backups used for rollback.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.commitUpgrade)))
```

Comment:
Why are these separate command `Action`s, more specifically, why are thy separate repo sessions (`directRepositoryWriteAction`)? Do they need to be?

Reply:
Good question, each of these functions are individual idempotent components which are retryable. Also, it is easier to read it in phases of intent -> drain -> upgrade -> commit.

We do not need this to be separate write sessions though, I can fix this in a follow-up PR.